instrumentation/apache-httpclient/apache-httpclient-4.3/library/src/main/java/io/opentelemetry/instrumentation/apachehttpclient/v4_3/TracingProtocolExec.java

```diff
@@ -0,0 +1,172 @@
+      HttpClientContext httpContext,
+      HttpExecutionAware httpExecutionAware)
+      throws IOException, HttpException {
+    Context context = httpContext.getAttribute(REQUEST_CONTEXT_ATTRIBUTE_ID, Context.class);
+    if (context != null) {
+      ApacheHttpClientRequest instrumenterRequest =
+          httpContext.getAttribute(REQUEST_WRAPPER_ATTRIBUTE_ID, ApacheHttpClientRequest.class);
+      // Request already had a context so a redirect. Don't create a new span just inject and
+      // execute.
+      propagators.getTextMapPropagator().inject(context, request, HttpHeaderSetter.INSTANCE);
+      return execute(route, request, instrumenterRequest, httpContext, httpExecutionAware, context);
+    }
+
+    HttpHost host = null;
+    if (route.getTargetHost() != null) {
+      host = route.getTargetHost();
+    } else if (httpContext.getTargetHost() != null) {
+      host = httpContext.getTargetHost();
+    }
+    if (host != null) {
+      if ((host.getSchemeName().equals("https") && host.getPort() == 443)
+          || (host.getSchemeName().equals("http") && host.getPort() == 80)) {
+        // port seems to be added to the host by route planning for standard ports even if not
+        // specified in the URL. There doesn't seem to be a way to differentiate between explicit
+        // and implicit port, but ignore in both cases to match the more common case.
+        host = new HttpHost(host.getHostName(), -1, host.getSchemeName());
+      }
+    }
+    ApacheHttpClientRequest instrumenterRequest = new ApacheHttpClientRequest(host, request);
+
+    Context parentContext = Context.current();
+    if (!instrumenter.shouldStart(parentContext, instrumenterRequest)) {
+      return exec.execute(route, request, httpContext, httpExecutionAware);
+    }
+
+    context = instrumenter.start(parentContext, instrumenterRequest);
+    httpContext.setAttribute(REQUEST_CONTEXT_ATTRIBUTE_ID, context);
+    httpContext.setAttribute(REQUEST_WRAPPER_ATTRIBUTE_ID, instrumenterRequest);
+    httpContext.setAttribute(REDIRECT_COUNT_ATTRIBUTE_ID, 0);
+
+    propagators.getTextMapPropagator().inject(context, request, HttpHeaderSetter.INSTANCE);
+
+    return execute(route, request, instrumenterRequest, httpContext, httpExecutionAware, context);
+  }
+
+  private CloseableHttpResponse execute(
+      HttpRoute route,
+      HttpRequestWrapper request,
+      ApacheHttpClientRequest instrumenterRequest,
+      HttpClientContext httpContext,
+      HttpExecutionAware httpExecutionAware,
+      Context context)
+      throws IOException, HttpException {
+    CloseableHttpResponse response = null;
+    Throwable error = null;
+    try (Scope ignored = context.makeCurrent()) {
+      response = exec.execute(route, request, httpContext, httpExecutionAware);
+      return response;
+    } catch (Throwable e) {
+      error = e;
+      throw e;
+    } finally {
+      if (!pendingRedirect(context, httpContext, request, instrumenterRequest, response)) {
```

Comment:
It is unfortunate that we can't wrap around redirect and automatic retry. Perhaps we should just let this instrumentation create multiple spans when there is a redirect? If I understand correctly if retry is enable then we only record the first failing request even when retry succeeds?

Reply:
Nah, the normal redirect tests in HttpClientTest are passing so we're recording it as a success. And the timing is around all the redirects since we end the span after the last one.

The code is sort of complex but I think the modeling is consistent with our other instrumentation so guess it's not too bad.

Reply:
Yeah redirects look ok. There is also an option to retry request when it fails, imo it would need similar handling as redirects to make sure that we don't end span on the first attempt. We don't have a test for that.

Reply:
@laurit We discussed a bit and figure it's good to try this code first and iterate later, since we don't have any tests for retries yet. We'd probably need to check everything in the future more broadly

Reply:
I just wanted to point out that with library instrumentation we have limited options to force the framework to do things as we like so we could instead create spans in a way that fits best with the integration points the framework provides us. This way there should be less corner cases we have to worry about. Sure it will be bad that different library instrumentations don't behave in a consistent way and even worse would be testing them, but instrumentation would be simpler and the chance of something breaking because http client is configured slightly differently or customized a bit should be smaller.